File: kotlin/everybody_codes/src/main/kotlin/org/gristle/pdxpuzzles/everybodycodes/y2024/Y24D14.kt
package org.gristle.pdxpuzzles.everybodycodes.y2024

import org.gristle.pdxpuzzles.everybodycodes.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.Xyz
import java.util.ArrayDeque

object Y24D14 : Day {
    override fun part1(input: String): Int = growBranch(input).maxOf { it.y }

    override fun part2(input: String): Int = input
        .lineSequence()
        .flatMap { line -> growBranch(line) }
        .toSet()
        .size

    override fun part3(input: String): Int {
        val branches: List<List<Xyz>> = input
            .lines()
            .map { line -> growBranch(line).toList() }
        val leaves = branches
            .map { branch -> branch.last() }
        val tree = branches.flatten().toSet()
        val treeHeight = tree.maxOf { it.y }
        val tapPoints = List(treeHeight) { IntArray(leaves.size) }
        val offsets = listOf(
            Xyz(1, 0, 0),
            Xyz(-1, 0, 0),
            Xyz(0, 1, 0),
            Xyz(0, -1, 0),
            Xyz(0, 0, 1),
            Xyz(0, 0, -1),
        )
        for ((leafIndex, leaf) in leaves.withIndex()) {
            var taps = 0
            val q = ArrayDeque<Pair<Xyz, Int>>()
            q.add(leaf to 0)
            val visited = mutableSetOf<Xyz>()
            while (q.isNotEmpty()) {
                val (pos, weight) = q.pop()

                // skip because already visited
                if (!visited.add(pos)) continue

                // trunk found
                if (pos.x == 0 && pos.z == 0) {
                    tapPoints[pos.y - 1][leafIndex] = weight

                    // end condition met
                    if (++taps == treeHeight) break
                }

                // get neighbors
                offsets
                    .map { pos + it }
                    .filter { it !in visited && it in tree }
                    .map { (x, y, z) -> Xyz(x, y, z) to weight + 1 }
                    .forEach { state -> q.add(state) }
            }
        }
        return tapPoints.map { it.sum() }.filter { it != 0 }.min()
    }

    private fun String.splitAt(index: Int): Pair<String, String> {
        require(index in 0..length) { "Index must be within the bounds of the string length." }
        return substring(0, index) to substring(index)
    }

    private fun growBranch(input: String) = input
        .splitToSequence(',')
        .flatMap { instruction ->
            val (dir, dist) = instruction.splitAt(1)
            val direction = dir[0]
            val distance = dist.toInt()
            generateSequence { direction }.take(distance)
        }.runningFold(Xyz(0, 0, 0)) { pos, direction ->
            pos + when (direction) {
                'U' -> Xyz(0, 1, 0)
                'D' -> Xyz(0, -1, 0)
                'L' -> Xyz(-1, 0, 0)
                'R' -> Xyz(1, 0, 0)
                'F' -> Xyz(0, 0, 1)
                'B' -> Xyz(0, 0,-1)
                else -> throw IllegalArgumentException("Unrecognized direction: $direction")
            }
        }.drop(1) // initial value for runningFold not part of branch

}

fun main() = Day.runDay(Y24D14::class)

//    Quest 1: 154 (15ms)
//    Quest 2: 4908 (25ms)
//    Quest 3: 1512 (248ms)
//    Total time: 289ms

File: rust/everybody-codes/src/bin/2024/y24d14.rs
use everybody_codes::utilities::inputs::get_event_inputs;
use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
use std::{
    collections::{HashSet, VecDeque},
    sync::LazyLock,
};
use utilities::structs::{
    coord::Coord3,
    stopwatch::{ReportDuration, Stopwatch},
};

static MOVES: LazyLock<[Coord3; 6]> = LazyLock::new(|| {
    [
        Coord3::new3d(1, 0, 0),
        Coord3::new3d(-1, 0, 0),
        Coord3::new3d(0, 1, 0),
        Coord3::new3d(0, -1, 0),
        Coord3::new3d(0, 0, 1),
        Coord3::new3d(0, 0, -1),
    ]
});
fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let (input1, input2, input3) = get_event_inputs(24, 14);
    println!("Inputs loaded ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input1), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input2), stopwatch.lap().report());
    println!("3. {} ({})", part3(&input3), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn part1(input: &str) -> usize {
    grow_branch(input)
        .max_by(|a, b| a.y().cmp(&b.y()))
        .unwrap()
        .y() as usize
}

fn part2(input: &str) -> usize {
    input
        .lines()
        .flat_map(|line| grow_branch(line))
        .collect::<HashSet<Coord3>>()
        .len()
}

fn part3(input: &str) -> usize {
    let branches: Vec<Vec<Coord3>> = input
        .lines()
        .map(|line| grow_branch(line).collect())
        .collect();
    let leaves: Vec<Coord3> = branches
        .iter()
        .map(|branch| *branch.last().unwrap())
        .collect();
    let tree: HashSet<Coord3> = branches
        .into_iter()
        .flat_map(|branch| branch.into_iter())
        .collect();
    let height = tree.iter().map(|segment| segment.y()).max().unwrap() as usize;

    let leaf_to_trunk: Vec<Vec<usize>> = leaves
        .par_iter()
        .map(|leaf| {
            let mut taps = 0;
            let mut tap_distance = vec![0; height];
            let mut q = VecDeque::new();
            q.push_back((*leaf, 0));
            let mut visited = HashSet::new();
            while let Some((pos, weight)) = q.pop_front() {
                if !visited.insert(pos) {
                    continue;
                }
                if pos.x() == 0 && pos.z() == 0 {
                    tap_distance[(pos.y() - 1) as usize] = weight;
                    taps += 1;
                    if taps == height {
                        break;
                    }
                }
                MOVES
                    .iter()
                    .map(|dir| *dir + pos)
                    .filter(|n_pos| !visited.contains(n_pos) && tree.contains(n_pos))
                    .map(|n_pos| (n_pos, weight + 1))
                    .for_each(|state| q.push_back(state))
            }
            tap_distance
        })
        .collect();

    let mut tap_points = vec![vec![0; leaves.len()]; height];

    for (leaf_index, trunk_counts) in leaf_to_trunk.iter().enumerate() {
        for (trunk_index, &count) in trunk_counts.iter().enumerate() {
            tap_points[trunk_index][leaf_index] = count;
        }
    }

    tap_points
        .iter()
        .map(|distances| distances.iter().sum::<usize>())
        .filter(|&murkiness| murkiness != 0)
        .min()
        .unwrap()
}

fn grow_branch<'a>(input: &'a str) -> impl Iterator<Item = Coord3> + 'a {
    input
        .split(',')
        .flat_map(|instruction| {
            let (dir, distance) = instruction.split_at(1);
            let dir = dir.as_bytes()[0] as char;
            let distance: usize = distance.parse().unwrap();
            vec![dir; distance].into_iter()
        })
        .scan(Coord3::origin(), |pos, dir| {
            *pos += match dir {
                'U' => Coord3::new3d(0, 1, 0),
                'D' => Coord3::new3d(0, -1, 0),
                'L' => Coord3::new3d(-1, 0, 0),
                'R' => Coord3::new3d(1, 0, 0),
                'F' => Coord3::new3d(0, 0, 1),
                'B' => Coord3::new3d(0, 0, -1),
                other => {
                    panic!("Unrecognized direction: {other}.");
                }
            };
            Some(*pos)
        })
}

#[test]
fn tests() {
    let tests = [
        "U5,R3,D2,L5,U4,R5,D2",
        "U5,R3,D2,L5,U4,R5,D2
U6,L1,D2,R3,U2,L1",
        "U5,R3,D2,L5,U4,R5,D2
U6,L1,D2,R3,U2,L1",
        "U20,L1,B1,L2,B1,R2,L1,F1,U1
U10,F1,B1,R1,L1,B1,L1,F1,R2,U1
U30,L2,F1,R1,B1,R1,F2,U1,F1
U25,R1,L2,B1,U1,R2,F1,L2
U16,L1,B1,L1,B3,L1,B1,F1",
    ];
    assert_eq!(7, part1(tests[0]));
    assert_eq!(32, part2(tests[1]));
    assert_eq!(5, part3(tests[2]));
    assert_eq!(46, part3(tests[3]));
}
